src/wp_chachapoly_aead.c

```diff
@@ -0,0 +1,824 @@
+/* Dispatch functions for wp_chacha20_poly1305 cipher */
+
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/evp.h>
+#include <wolfprovider/settings.h>
+#include <wolfprovider/alg_funcs.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/chacha20_poly1305.h>
+
+#ifdef WP_HAVE_CHACHAPOLY
+
+#define CHACHA20_POLY1305_AEAD_INITIAL_COUNTER  0
+
+#define CHACHA_U8TOU32(p)  ( \
+    ((unsigned int)(p)[0])     | ((unsigned int)(p)[1]<<8) | \
+    ((unsigned int)(p)[2]<<16) | ((unsigned int)(p)[3]<<24)  )
+
+#define POLY1305_BLOCK_SIZE  16
+#define CHACHA_CTR_SIZE 16
+
+/** 
+ * Authenticated Encryption with Associated Data structure.
+ */
+typedef struct wp_CP_AeadCtx {
```

Comment:
How much is in common with wp_aes_aead.c?
If a lot then consider merging and renaming it wp_aead.c.